x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/score/Decay.java

```diff
@@ -0,0 +1,618 @@
+                valueFactory,
+                (Integer) originFolded,
+                (Integer) scaleFolded,
+                (Integer) offsetFolded,
+                decayFolded,
+                typeFolded
+            );
+            case DOUBLE -> new DecayDoubleEvaluator.Factory(
+                source(),
+                valueFactory,
+                (Double) originFolded,
+                (Double) scaleFolded,
+                (Double) offsetFolded,
+                decayFolded,
+                typeFolded
+            );
+            case LONG -> new DecayLongEvaluator.Factory(
+                source(),
+                valueFactory,
+                (Long) originFolded,
+                (Long) scaleFolded,
+                (Long) offsetFolded,
+                decayFolded,
+                typeFolded
+            );
+            case GEO_POINT -> new DecayGeoPointEvaluator.Factory(
+                source(),
+                valueFactory,
+                (BytesRef) originFolded,
+                (BytesRef) scaleFolded,
+                (BytesRef) offsetFolded,
+                decayFolded,
+                typeFolded
+            );
+            case CARTESIAN_POINT -> new DecayCartesianPointEvaluator.Factory(
+                source(),
+                valueFactory,
+                (BytesRef) originFolded,
+                (Double) scaleFolded,
+                (Double) offsetFolded,
+                decayFolded,
+                typeFolded
+            );
+            case DATETIME -> new DecayDatetimeEvaluator.Factory(
+                source(),
+                valueFactory,
+                (Long) originFolded,
+                (Long) scaleFolded,
+                (Long) offsetFolded,
+                decayFolded,
+                typeFolded
+            );
+            case DATE_NANOS -> new DecayDateNanosEvaluator.Factory(
+                source(),
+                valueFactory,
+                (Long) originFolded,
+                (Long) scaleFolded,
+                (Long) offsetFolded,
+                decayFolded,
+                typeFolded
+            );
+            default -> throw new UnsupportedOperationException("Unsupported data typeExpr: " + valueDataType);
+        };
+    }
+
+    @Override
+    public BiConsumer<LogicalPlan, Failures> postAnalysisPlanVerification() {
+        return (LogicalPlan plan, Failures failures) -> {
+            Expression offset = (Expression) resolvedOptions.get(OFFSET);
+            Expression decay = (Expression) resolvedOptions.get(DECAY);
+            Expression type = (Expression) resolvedOptions.get(TYPE);
+
+            Map<String, Expression> constantExpressions = new HashMap<>();
+            constantExpressions.put(OFFSET, offset);
+            constantExpressions.put(DECAY, decay);
+            constantExpressions.put(TYPE, type);
+
+            // Verify options are constant foldable
+            constantExpressions.forEach((exprName, expr) -> {
```

Comment:
I believe there's no way they can't be constant foldable from what we discussed, right?